File: frontend/src/game/upgrades/stat_modifiers/damage_2.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Damage2Def: UpgradeDef = {
  id: "damage_2",
  name: "Devastation",
  description: "+0.8% damage.",
  rarity: RarityID.Uncommon,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 6,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 0.008,
  isMultiplier: true,
  stackable: true,
  maxStacks: 99999,
}

export class Damage2 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/damage_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Damage2(UpgradeImplementation):
    id = 'damage_2'
    name = 'Devastation'
    description = '+0.8% damage.'
    rarity = 'uncommon'
    type = 'stat_modifier'
    cost = 6

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.008
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
